builder/src/lib.rs

```diff
@@ -237,6 +237,11 @@ fn finalize_blob(
         // blob file.
         if !is_tarfs {
             blob_writer.finalize(Some(blob_meta_id))?;
+            if let Some(blob_cache) = ctx.blob_cache_generator.as_ref() {
```

Comment:
How about moving the code piece out of `!is_tarfs`?
We should finalize the storage if ctx.blob_cache_generator is valid.